src/vs/editor/contrib/unicodeHighlighter/unicodeHighlighter.ts

```diff
@@ -0,0 +1,504 @@
+			&& this.range.startColumn <= anchor.range.startColumn
+			&& this.range.endColumn >= anchor.range.endColumn
+		);
+	}
+}
+
+export class UnicodeHighlighterHoverParticipant implements IEditorHoverParticipant<MarkdownHover> {
+
+	constructor(
+		private readonly _editor: ICodeEditor,
+		private readonly _hover: IEditorHover,
+		@IModeService private readonly _modeService: IModeService,
+		@IOpenerService private readonly _openerService: IOpenerService,
+	) {
+	}
+
+	computeSync(anchor: HoverAnchor, lineDecorations: IModelDecoration[]): MarkdownHover[] {
+		if (!this._editor.hasModel() || anchor.type !== HoverAnchorType.Range) {
+			return [];
+		}
+
+		const model = this._editor.getModel();
+
+		const unicodeHighlighter = this._editor.getContribution<UnicodeHighlighter>(UnicodeHighlighter.ID);
+
+
+		const result: MarkdownHover[] = [];
+		for (const d of lineDecorations) {
+
+			const highlightInfo = unicodeHighlighter.getDecorationInfo(d.id);
+			if (!highlightInfo) {
+				continue;
+			}
+			const char = model.getValueInRange(d.range);
+			// text refers to a single character.
+			const codePoint = char.codePointAt(0)!;
+
+			function formatCodePoint(codePoint: number) {
+				return `\`U+${codePoint.toString(16)}\` "${`${escapedMarkdownInlineCode(String.fromCodePoint(codePoint))}`}"`;
+			}
+
+			const codePointStr = formatCodePoint(codePoint);
+
+			let reason: string;
+			switch (highlightInfo.reason.kind) {
+				case 'ambiguous':
+					reason = nls.localize(
+						'unicodeHighlight.characterIsAmbiguous',
+						'The character {0} could be confused with the more common character {1}.',
+						codePointStr,
+						formatCodePoint(highlightInfo.reason.confusableWith.codePointAt(0)!)
+					);
+					break;
+
+				case 'invisible':
+					reason = nls.localize(
+						'unicodeHighlight.characterIsInvisible',
+						'The character {0} is invisible.',
+						codePointStr
+					);
+					break;
+
+				case 'nonBasicAscii':
+					reason = nls.localize(
+						'unicodeHighlight.characterIsNonBasicAscii',
+						'The character {0} is not a basic ASCII character.',
+						codePoint
+					);
+					break;
+			}
+
+			const adjustSettingsArgs: ShowExcludeOptionsArgs = {
+				codePoint: codePoint,
+				reason: highlightInfo.reason.kind,
+			};
+
+			const adjustSettings = nls.localize('unicodeHighlight.adjustSettings', 'Adjust settings');
+			const contents: Array<IMarkdownString> = [{
+				value: `${reason} [${adjustSettings}](command:${ShowExcludeOptions.ID}?${encodeURIComponent(JSON.stringify(adjustSettingsArgs))})`,
+				isTrusted: true,
+			}];
+
+			result.push(new MarkdownHover(this, d.range, contents));
+		}
+		return result;
+	}
+
+	public renderHoverParts(hoverParts: MarkdownHover[], fragment: DocumentFragment, statusBar: IEditorHoverStatusBar): IDisposable {
+		return renderMarkdownHovers(hoverParts, fragment, this._editor, this._hover, this._modeService, this._openerService);
+	}
+}
+
+function escapedMarkdownInlineCode(code: string): string {
```

Comment:
This function appears overengineered to me. It looks like it is only called with a single code point as the value of `code`, yet its implementation deals with the complex case of escaping arbitrary long sequences of backticks. A more simple function could be changed to simply cater for `` code === '`' `` and hard code the return value of ``` `` ` `` ``` in that case.

```
function renderCodePointAsInlineCode(codePoint: number): string {
	if (codePoint === CharCode.BackTick) {
		return '`` ` ``';
	}
	return '`' + String.fromCodePoint(codePoint) + '`';
}
```